File: src/Yass/List.hs
{-# LANGUAGE OverloadedStrings #-}

module Yass.List (runList, normalizeDescription, truncateDescription) where

import Control.Exception (IOException, try)
import Data.IORef
import Data.List (sortBy)
import Data.Ord (comparing)
import qualified Data.Set as Set
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import qualified Data.Text.Normalize as Normalize
import System.Directory (getCurrentDirectory, makeAbsolute)
import System.Environment (lookupEnv)
import System.Exit (ExitCode(..))
import System.FilePath (normalise)
import System.IO (hFlush, hIsTerminalDevice, stderr, stdout)
import System.Process (readProcess)

import Yass.Types hiding (ExitSuccess)
import Yass.ErrorLine (formatErrorLineNoLine, relativizePath)
import Yass.Parser (RawDocument(..), RawValue(..), ParseResult(..), parseYassFile)
import Yass.ProjectRoot (findProjectRoot)
import Yass.Discover (discoverSpecFiles, isYassYamlFile)
import Yass.Glob (expandGlob, hasGlobChars)

-- | Run the list subcommand.
-- Takes argv (subcommand args, NOT including "list" itself).
runList :: [String] -> IO ExitCode
runList args = do
  cwd <- getCurrentDirectory

  -- Check for colon in paths
  let colonPaths = filter (':' `elem`) args
  case colonPaths of
    (p:_) -> do
      emitErr cwd "yass" PathColonInPath
        ("path contains an unsupported colon character: " <> T.pack p)
      return (ExitFailure 2)
    [] -> do
      -- Find project root
      rootResult <- findProjectRoot cwd
      case rootResult of
        Left _ -> do
          emitErr cwd (relativizePath cwd cwd) FindrootNoMarker
            "no project root marker found"
          return (ExitFailure 2)
        Right projectRoot -> do
          -- Resolve input files (supports multiple args, globs, dirs)
          filesResult <- resolveInputFiles cwd projectRoot args
          case filesResult of
            Left err -> do
              emitYassErr cwd err
              return (ExitFailure 2)
            Right files
              | null files -> return ExitSuccess  -- no files found → exit 0, no output
              | otherwise -> do
                  -- Deduplicate by absolute path, then sort by NFC
                  deduped <- deduplicateFiles files
                  let sorted = sortByNFC deduped

                  -- Get terminal width if stdout is a TTY
                  isTTY <- hIsTerminalDevice stdout
                  termWidth <- if isTTY then getTerminalWidth else return Nothing

                  -- Process each file
                  hadParseError <- processFiles cwd sorted termWidth
                  hFlush stdout
                  if hadParseError
                    then return (ExitFailure 1)
                    else return ExitSuccess

-- | Resolve input files from arguments (handles globs, dirs, explicit files).
resolveInputFiles :: FilePath -> FilePath -> [String] -> IO (Either YassError [FilePath])
resolveInputFiles cwd projectRoot [] = do
  result <- discoverSpecFiles cwd projectRoot Nothing
  case result of
    Left errs -> return $ Left (head errs)
    Right fs  -> return $ Right fs
resolveInputFiles cwd projectRoot allArgs = do
  allFiles <- newIORef ([] :: [FilePath])
  let processArg arg
        | hasGlobChars arg = do
            globResult <- expandGlob arg
            case globResult of
              Left msg -> return $ Just $ YassError
                { yeFile = Just "yass", yeLine = Nothing
                , yeCode = GlobNoMatch, yeMessage = T.pack msg }
              Right paths -> do
                -- For glob-expanded paths, keep only .yass.yaml files
                let yassFiles = filter isYassYamlFile paths
                modifyIORef' allFiles (++ yassFiles)
                return Nothing
        | otherwise = do
            result <- discoverSpecFiles cwd projectRoot (Just arg)
            case result of
              Left errs -> return $ Just (head errs)
              Right fs -> do
                modifyIORef' allFiles (++ fs)
                return Nothing

  errs <- mapM processArg allArgs
  case [e | Just e <- errs] of
    (e:_) -> return $ Left e
    [] -> do
      fs <- readIORef allFiles
      return $ Right fs

-- | Process files and emit list rows. Returns True if any file failed to parse.
processFiles :: FilePath -> [FilePath] -> Maybe Int -> IO Bool
processFiles cwd files termWidth = do
  results <- mapM (processFile cwd termWidth) files
  return (or results)

-- | Process a single file. Returns True if parse failed.
processFile :: FilePath -> Maybe Int -> FilePath -> IO Bool
processFile cwd termWidth fp = do
  result <- parseYassFile fp
  case result of
    ParseError _ _ -> do
      let errLine = formatErrorLineNoLine cwd fp YamlMalformed
                      "YAML well-formedness error"
      TIO.hPutStrLn stderr errLine
      hFlush stderr
      return True
    ParseOK docs -> do
      let relPath = relativizePath cwd fp
      case docs of
        [] -> return False
        (preamble:specDocs) -> do
          let desc = extractDescription preamble
              normDesc = normalizeDescription desc
          mapM_ (emitListRow relPath normDesc termWidth) (extractSpecNames specDocs)
          return False

-- | Extract spec names from non-preamble documents, preserving doc order.
extractSpecNames :: [RawDocument] -> [T.Text]
extractSpecNames = concatMap getSpecName
  where
    getSpecName doc = case lookup3 "spec" (rawDocEntries doc) of
      Just (RVString name, _) -> [name]
      _ -> []

-- | Extract the preamble description.
extractDescription :: RawDocument -> T.Text
extractDescription doc = case lookup3 "description" (rawDocEntries doc) of
  Just (RVString desc, _) -> desc
  _ -> ""

-- | Normalize a description: replace all whitespace runs with single space,
-- strip leading/trailing whitespace.
normalizeDescription :: T.Text -> T.Text
normalizeDescription t
  | T.null stripped = ""
  | otherwise       = T.intercalate " " (T.words stripped)
  where
    stripped = T.strip t

-- | Truncate a description to fit within a terminal width.
--
-- @truncateDescription fileLen nameLen width desc@
--
-- The line format is: file\\tname\\tdescription
-- Prefix width = fileLen + 1 (tab) + nameLen + 1 (tab) = fileLen + nameLen + 2
-- Marker = "..." (3 chars)
--
-- * If description is empty -> empty, no marker.
-- * If prefix + marker >= width -> empty, no marker.
-- * If description fits -> return as-is.
-- * Otherwise -> truncate + "...".
truncateDescription :: Int -> Int -> Int -> T.Text -> T.Text
truncateDescription fileLen nameLen width desc
  | T.null desc = ""
  | prefixLen + markerLen >= width = ""
  | T.length desc <= available = desc
  | truncLen <= 0 = ""
  | otherwise = T.take truncLen desc <> "..."
  where
    prefixLen = fileLen + nameLen + 2  -- file + TAB + name + TAB
    markerLen = 3  -- "..."
    available = width - prefixLen
    truncLen  = available - markerLen

-- | Emit a list row for a spec.
emitListRow :: FilePath -> T.Text -> Maybe Int -> T.Text -> IO ()
emitListRow filePath desc termWidth specName = do
  let -- Replace literal tab characters in the file-path field with spaces
      fp = T.replace "\t" " " (T.pack filePath)
      -- Apply NFC normalization to the description before emission
      nfcDesc = Normalize.normalize Normalize.NFC desc
  case termWidth of
    Nothing ->
      -- Not a TTY: full output, no truncation
      TIO.putStrLn $ fp <> "\t" <> specName <> "\t" <> nfcDesc
    Just width -> do
      -- TTY: truncate description to fit width
      let truncated = truncateDescription (T.length fp) (T.length specName) width nfcDesc
      TIO.putStrLn $ fp <> "\t" <> specName <> "\t" <> truncated

-- | Get terminal width from COLUMNS env var or terminal query.
getTerminalWidth :: IO (Maybe Int)
getTerminalWidth = do
  colsEnv <- lookupEnv "COLUMNS"
  case colsEnv of
    Just s -> case reads s of
      [(n, "")] | n > 0 -> return (Just n)
      _ -> queryTerminalWidth
    Nothing -> queryTerminalWidth

-- | Query terminal width from OS using tput. Falls back to Just 80 on failure.
queryTerminalWidth :: IO (Maybe Int)
queryTerminalWidth = do
  result <- try (readProcess "tput" ["cols"] "") :: IO (Either IOException String)
  case result of
    Right s -> case reads (strip s) of
      [(n, "")] | n > 0 -> return (Just n)
      _                 -> return (Just 80)
    Left _ -> return (Just 80)
  where
    strip = reverse . dropWhile isSpace . reverse . dropWhile isSpace
    isSpace c = c == ' ' || c == '\n' || c == '\r' || c == '\t'

-- | Deduplicate files by normalized absolute path.
deduplicateFiles :: [FilePath] -> IO [FilePath]
deduplicateFiles files = do
  absFiles <- mapM (\f -> do
    abs' <- makeAbsolute f
    return (normalise abs', f)
    ) files
  let go _ [] = []
      go s ((absP, origP):rest)
        | Set.member absP s = go s rest
        | otherwise = origP : go (Set.insert absP s) rest
  return $ go (Set.empty :: Set.Set FilePath) absFiles

-- | Sort file paths by NFC-normalized Unicode code-point order.
sortByNFC :: [FilePath] -> [FilePath]
sortByNFC = sortBy (comparing nfcNorm)
  where nfcNorm = Normalize.normalize Normalize.NFC . T.pack

-- | Emit an error to stderr.
emitErr :: FilePath -> FilePath -> ErrorCode -> T.Text -> IO ()
emitErr cwd fp code msg = do
  TIO.hPutStrLn stderr $ formatErrorLineNoLine cwd fp code msg
  hFlush stderr

-- | Emit a YassError to stderr.
emitYassErr :: FilePath -> YassError -> IO ()
emitYassErr cwd err = do
  let fp = maybe "yass" id (yeFile err)
  TIO.hPutStrLn stderr $ formatErrorLineNoLine cwd fp (yeCode err) (yeMessage err)
  hFlush stderr

-- | Lookup helper for (key, value, line) triples.
lookup3 :: T.Text -> [(T.Text, RawValue, Int)] -> Maybe (RawValue, Int)
lookup3 _ [] = Nothing
lookup3 k ((k', v, ln):rest)
  | k == k'   = Just (v, ln)
  | otherwise  = lookup3 k rest

File: internal/list/list.go
package list

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"sort"
	"strings"

	"github.com/rivo/uniseg"
	"golang.org/x/text/unicode/norm"
	"gopkg.in/yaml.v3"

	yerrors "github.com/shakefu/yass/internal/errors"
	"github.com/shakefu/yass/internal/exitcode"
	"github.com/shakefu/yass/internal/parser"
	"github.com/shakefu/yass/internal/shared"
)

// whitespaceRun matches one or more whitespace characters (space, tab, newline, etc.).
var whitespaceRun = regexp.MustCompile(`\s+`)

// Run executes the list subcommand.
// args = positional arguments after "list".
// Returns an exit code: 0 success, 1 parse errors, 2 usage/path errors.
func Run(args []string, stdout io.Writer, stderr io.Writer, isTTY bool, termWidth int) int {
	cwd, err := os.Getwd()
	if err != nil {
		line := yerrors.FormatError("yass", 0, yerrors.CodeInternalUncaught, fmt.Sprintf("internal error: %s", err))
		fmt.Fprint(stderr, line)
		return exitcode.Processing
	}

	// Determine optional input path from args.
	var inputPath string
	if len(args) > 0 {
		inputPath = args[0]
	}

	// Check for colon in path argument.
	// ErrorLine spec: MUST-NOT emit any input path containing ":".
	if inputPath != "" && strings.Contains(inputPath, ":") {
		line := yerrors.FormatError(
			"yass",
			0,
			yerrors.CodePathColonInPath,
			fmt.Sprintf(yerrors.Messages[yerrors.CodePathColonInPath], inputPath),
		)
		fmt.Fprint(stderr, line)
		return exitcode.Usage
	}

	// When no path given, find project root and discover from there.
	var projectRoot string
	if inputPath == "" {
		root, err := shared.FindProjectRoot(cwd)
		if err != nil {
			if pe, ok := err.(*shared.PathError); ok {
				line := yerrors.FormatError(
					yerrors.FormatPath(cwd, cwd),
					0,
					pe.Code,
					pe.Message,
				)
				fmt.Fprint(stderr, line)
			} else {
				line := yerrors.FormatError("yass", 0, yerrors.CodeInternalUncaught, fmt.Sprintf("internal error: %s", err))
				fmt.Fprint(stderr, line)
			}
			return exitcode.Usage
		}
		projectRoot = root
	}

	// Discover spec files.
	filePaths, nonFatalErrs, discoverErr := shared.DiscoverSpecFiles(inputPath, cwd, projectRoot)
	if discoverErr != nil {
		emitPathError(discoverErr, cwd, stderr)
		return exitcode.Usage
	}
	for _, e := range nonFatalErrs {
		fmt.Fprint(stderr, e)
	}

	// No files found → exit 0, no output.
	if len(filePaths) == 0 {
		return exitcode.Success
	}

	// Sort files by Unicode code-point order on NFC-normalized path.
	sort.Slice(filePaths, func(i, j int) bool {
		ni := norm.NFC.String(filePaths[i])
		nj := norm.NFC.String(filePaths[j])
		return ni < nj
	})

	hadParseError := false

	for _, filePath := range filePaths {
		rows, parseErr := processFile(filePath, cwd)
		if parseErr != nil {
			fmt.Fprint(stderr, parseErr.Error())
			hadParseError = true
			continue
		}
		for _, row := range rows {
			emitRow(stdout, row, isTTY, termWidth)
		}
	}

	if hadParseError {
		return exitcode.Processing
	}
	return exitcode.Success
}

// listRow holds one output row before formatting.
type listRow struct {
	FilePath    string
	SpecName    string
	Description string
}

// processFile parses a single file and returns list rows.
// Returns a formatted error string on parse failure.
func processFile(filePath, cwd string) ([]listRow, *parseErrorLine) {
	result, err := parser.ParseFile(filePath)
	if err != nil {
		formattedPath := yerrors.FormatPath(filePath, cwd)
		var line int
		var code, msg string
		if pe, ok := err.(*parser.ParseError); ok {
			line = pe.Line
			code = pe.Code
			msg = pe.Message
		} else {
			code = yerrors.CodeYamlMalformed
			msg = yerrors.Messages[yerrors.CodeYamlMalformed]
		}
		errLine := yerrors.FormatError(formattedPath, line, code, msg)
		return nil, &parseErrorLine{line: errLine}
	}

	if len(result.Documents) == 0 {
		return nil, nil
	}

	// Extract preamble description from the first document without "spec" key.
	var description string
	var specDocs []*yaml.Node
	startIdx := 0

	// Check if first doc is a preamble (no "spec" key).
	if len(result.Documents) > 0 {
		firstDoc := result.Documents[0]
		if !docHasSpecKey(firstDoc) {
			description = extractDescription(firstDoc)
			startIdx = 1
		}
	}

	// Collect spec documents.
	for i := startIdx; i < len(result.Documents); i++ {
		doc := result.Documents[i]
		if docHasSpecKey(doc) {
			specDocs = append(specDocs, doc)
		}
	}

	if len(specDocs) == 0 {
		return nil, nil
	}

	// Normalize file path: replace literal tab with space.
	displayPath := strings.ReplaceAll(filePath, "\t", " ")

	// Normalize description.
	normalizedDesc := normalizeDescription(description)

	var rows []listRow
	for _, doc := range specDocs {
		specName := extractSpecName(doc)
		rows = append(rows, listRow{
			FilePath:    displayPath,
			SpecName:    specName,
			Description: normalizedDesc,
		})
	}

	return rows, nil
}

// parseErrorLine wraps a formatted error string.
type parseErrorLine struct {
	line string
}

func (e *parseErrorLine) Error() string {
	return e.line
}

// docHasSpecKey checks if a YAML document node has a top-level "spec" key.
func docHasSpecKey(doc *yaml.Node) bool {
	if doc == nil {
		return false
	}
	// Document node wraps the actual content.
	node := doc
	if node.Kind == yaml.DocumentNode && len(node.Content) > 0 {
		node = node.Content[0]
	}
	if node.Kind != yaml.MappingNode {
		return false
	}
	for i := 0; i+1 < len(node.Content); i += 2 {
		key := node.Content[i]
		if key.Kind == yaml.ScalarNode && key.Value == "spec" {
			return true
		}
	}
	return false
}

// extractDescription extracts the "description" value from a YAML document node.
// Returns empty string if not found or not a string.
func extractDescription(doc *yaml.Node) string {
	if doc == nil {
		return ""
	}
	node := doc
	if node.Kind == yaml.DocumentNode && len(node.Content) > 0 {
		node = node.Content[0]
	}
	if node.Kind != yaml.MappingNode {
		return ""
	}
	for i := 0; i+1 < len(node.Content); i += 2 {
		key := node.Content[i]
		val := node.Content[i+1]
		if key.Kind == yaml.ScalarNode && key.Value == "description" {
			if val.Kind == yaml.ScalarNode && (val.Tag == "!!str" || val.Tag == "") {
				return val.Value
			}
			return ""
		}
	}
	return ""
}

// extractSpecName extracts the "spec" value from a YAML document node.
func extractSpecName(doc *yaml.Node) string {
	if doc == nil {
		return ""
	}
	node := doc
	if node.Kind == yaml.DocumentNode && len(node.Content) > 0 {
		node = node.Content[0]
	}
	if node.Kind != yaml.MappingNode {
		return ""
	}
	for i := 0; i+1 < len(node.Content); i += 2 {
		key := node.Content[i]
		val := node.Content[i+1]
		if key.Kind == yaml.ScalarNode && key.Value == "spec" {
			if val.Kind == yaml.ScalarNode {
				return val.Value
			}
			return ""
		}
	}
	return ""
}

// normalizeDescription normalizes a description string:
// - Replace all whitespace runs (newlines, tabs, spaces) with single ASCII space
// - Strip leading/trailing whitespace
// - NFC-normalize
func normalizeDescription(desc string) string {
	if desc == "" {
		return ""
	}
	// Replace all whitespace runs with a single space.
	s := whitespaceRun.ReplaceAllString(desc, " ")
	// Strip leading/trailing whitespace.
	s = strings.TrimSpace(s)
	// NFC-normalize.
	s = norm.NFC.String(s)
	return s
}

// graphemeLen returns the number of grapheme clusters in a string.
func graphemeLen(s string) int {
	count := 0
	g := uniseg.NewGraphemes(s)
	for g.Next() {
		count++
	}
	return count
}

// truncateToGraphemes truncates a string to at most maxGraphemes grapheme clusters.
// Returns the truncated string and whether it was actually shortened.
func truncateToGraphemes(s string, maxGraphemes int) (string, bool) {
	if maxGraphemes <= 0 {
		return "", s != ""
	}
	count := 0
	g := uniseg.NewGraphemes(s)
	var result strings.Builder
	for g.Next() {
		if count >= maxGraphemes {
			return result.String(), true
		}
		result.WriteString(g.Str())
		count++
	}
	return result.String(), false
}

// emitRow writes a single list row to stdout.
func emitRow(w io.Writer, row listRow, isTTY bool, termWidth int) {
	desc := row.Description

	if isTTY && termWidth > 0 {
		desc = truncateDescription(row.FilePath, row.SpecName, desc, termWidth)
	}

	fmt.Fprintf(w, "%s\t%s\t%s\n", row.FilePath, row.SpecName, desc)
}

// truncateDescription handles TTY truncation of the description field.
func truncateDescription(filePath, specName, desc string, width int) string {
	if desc == "" {
		return ""
	}

	pathLen := graphemeLen(filePath)
	nameLen := graphemeLen(specName)

	// Fixed columns: path + tab + name + tab = pathLen + 1 + nameLen + 1
	fixedCols := pathLen + 1 + nameLen + 1
	markerLen := 3 // "..."

	// If path+name+separators+marker >= width → empty third field, no marker
	if fixedCols+markerLen >= width {
		return ""
	}

	// Available space for description
	available := width - fixedCols
	descLen := graphemeLen(desc)

	if descLen <= available {
		// Fits without truncation.
		return desc
	}

	// Need to truncate. Available space for text = available - markerLen
	textSpace := available - markerLen
	if textSpace <= 0 {
		return ""
	}

	truncated, _ := truncateToGraphemes(desc, textSpace)
	return truncated + "..."
}

// emitPathError emits a path error to stderr.
func emitPathError(err error, cwd string, stderr io.Writer) {
	if pe, ok := err.(*shared.PathError); ok {
		formattedPath := yerrors.FormatPath(pe.Path, cwd)
		line := yerrors.FormatError(formattedPath, 0, pe.Code, pe.Message)
		fmt.Fprint(stderr, line)
	} else {
		line := yerrors.FormatError("yass", 0, yerrors.CodeInternalUncaught, fmt.Sprintf("internal error: %s", err))
		fmt.Fprint(stderr, line)
	}
}
